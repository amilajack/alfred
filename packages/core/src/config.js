/* eslint import/no-dynamic-require: off, no-param-reassign: off */
import path from 'path';
import fs from 'fs';
import formatPkg from 'format-package';
import mergeConfigs from '@alfred/merge-configs';
import Validate from './validation';

export type AlfredConfig = {
  extends?: Array<string> | string,
  npmClient: 'npm' | 'yarn',
  skills: Array<string | [string, Object]>,
  root: string,
  showConfigs: boolean
};

export function requireConfig(configName: string): any {
  try {
    // $FlowFixMe
    return require(`alfred-config-${configName}`);
  } catch (e) {
    try {
      // $FlowFixMe
      return require(configName);
    } catch (_e) {
      throw new Error(
        `Could not resolve "${configName}" module or "eslint-config-${configName}" module`
      );
    }
  }
}

export function getConfigs(config: AlfredConfig = {}): AlfredConfig {
  if (!config.extends) return config;
  if (!Array.isArray(config.extends) && typeof config.extends !== 'string') {
    throw new Error('.extends property must be an Array or a string');
  }
  // Convert extends: 'my-config' to extends: ['my-config']
  if (typeof config.extends === 'string') {
    config.extends = [config.extends];
  }
  // If a config is a string, require it
  const normalizedConfigs = config.extends.map(_config =>
    // $FlowFixMe
    typeof _config === 'string' ? requireConfig(_config) : _config
  );
  // If nothing to extend then return the config itself without the extends
  // property
  if (config.extends && !config.extends.length) {
    const newConfig = { ...config };
    delete newConfig.extends;
    return newConfig;
  }

  for (let i = 0; i < normalizedConfigs.length; i += 1) {
    // eslint-disable-next-line no-param-reassign
    normalizedConfigs[i] = getConfigs(normalizedConfigs[i]);
  }

  const mergedConfig = mergeConfigs(
    {},
    ...normalizedConfigs.map(e => e),
    config
  );
  delete mergedConfig.extends;
  return mergedConfig;
}

function validate(config: AlfredConfig): AlfredConfig {
  if (!config.extends) return config;
  // Validate if each config in `.extends` is a string
  if (Array.isArray(config.extends)) {
    config.extends.forEach(_config => {
      if (typeof _config !== 'string') {
        throw new Error(
          `Values in ".extends" property in Alfred config must be a string. Instead passed ${JSON.stringify(
            config.extends
          )}`
        );
      }
    });
  } else if (typeof config.extends !== 'string') {
    throw new Error(
      `Values in ".extends" property in Alfred config must be a string. Instead passed ${JSON.stringify(
        config.extends
      )}`
    );
  }
  return config;
}

export type ConfigSkillType = [string, any] | string;
type ConfigMap = Map<string, any>;

/**
 * @TODO @REFACTOR This function is not used outside of this module. Consider
 *                 not using it as the default export
 */
export default function Config(config: AlfredConfig): AlfredConfig {
  validate(config);

  const mergedConfig = getConfigs(config);
  if (!mergedConfig.skills || !mergedConfig.skills.length) return mergedConfig;

  const skillsMap: ConfigMap = new Map();
  const mappedSkills: ConfigMap = mergedConfig.skills.reduce(
    (map: ConfigMap, skill: ConfigSkillType) => {
      if (typeof skill === 'string') {
        map.set(skill, {});
        return map;
      }
      if (Array.isArray(skill)) {
        const [skillName, skillConfig] = skill;
        if (map.has(skillName)) {
          map.set(skillName, mergeConfigs({}, map.get(skillName), skillConfig));
        } else {
          map.set(skillName, skillConfig);
        }
        return map;
      }
      throw new Error(`Config type not supported: ${JSON.stringify(skill)}`);
    },
    skillsMap
  );

  mergedConfig.skills = Array.from(mappedSkills.entries());

  return mergedConfig;
}

const PKG_SORT_ORDER = [
  'name',
  'version',
  'private',
  'description',
  'keywords',
  'homepage',
  'bugs',
  'repository',
  'license',
  'author',
  'contributors',
  'files',
  'sideEffects',
  'main',
  'module',
  'jsnext:main',
  'browser',
  'types',
  'typings',
  'style',
  'example',
  'examplestyle',
  'assets',
  'bin',
  'man',
  'directories',
  'workspaces',
  'scripts',
  'dependencies',
  'devDependencies',
  'peerDependencies',
  'bundledDependencies',
  'bundleDependencies',
  'optionalDependencies',
  'resolutions',
  'engines',
  'engineStrict',
  'os',
  'cpu',
  'preferGlobal',
  'publishConfig',
  'betterScripts',
  'husky',
  'pre-commit',
  'commitlint',
  'lint-staged',
  'config',
  'nodemonConfig',
  'browserify',
  'babel',
  'browserslist',
  'xo',
  'prettier',
  'eslintConfig',
  'eslintIgnore',
  'stylelint',
  'jest',
  '...rest'
];

export type Pkg = { [x: string]: string };

export function formatPkgJson(pkg: Pkg): Promise<string> {
  return formatPkg(pkg, { order: PKG_SORT_ORDER });
}

/**
 * Writes an Alfred config to a user's package.json
 * @REFACTOR This should focus on only writing the alfred config, not the whole
 *           package.json
 */
export async function writeConfig(pkgPath: string, pkg: Pkg): AlfredConfig {
  const formattedPkg = await formatPkgJson(pkg);
  await fs.promises.writeFile(pkgPath, formattedPkg);
  return formattedPkg;
}

/**
 * @TODO @REFACTOR Make this the default exported function and rename it to Configs
 */
export async function loadConfig(
  projectRoot: string,
  pkgPath: string = path.join(projectRoot, 'package.json')
): Promise<{ pkg: Pkg, pkgPath: string, alfredConfig: AlfredConfig }> {
  if (!fs.existsSync(pkgPath)) {
    throw new Error('Current working directory does not have "package.json"');
  }

  // Read the package.json and validate the Alfred config
  const pkg = JSON.parse((await fs.promises.readFile(pkgPath)).toString());
  const rawAlfredConfig = pkg.alfred || {};
  Validate(rawAlfredConfig || {});
  // Format the config
  await writeConfig(pkgPath, pkg);

  const defaultOpts = {
    npmClient: 'npm',
    skills: [],
    root: projectRoot
  };

  const alfredConfig = Config({ ...defaultOpts, ...rawAlfredConfig });

  return { pkg, pkgPath, alfredConfig };
}
