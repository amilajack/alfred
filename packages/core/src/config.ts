/* eslint import/no-dynamic-require: off, no-param-reassign: off */
import * as path from 'path';
import * as fs from 'fs';
import mergeConfigs from '@alfred/merge-configs';
import { requireConfig } from '@alfred/helpers';
import ValidateConfig from './validation';
import Project, { formatPkgJson } from './project';
import CTF, {addMissingStdSkillsToCtf} from './ctf';
import { ConfigInterface, Skill, NpmClients, UnresolvedConfigInterface, ResolvedConfigInterface, CtfNode, CtfMap, InterfaceState } from './types';

type ConfigSkill = [string, any] | string;

type ConfigMap = Map<string, any>;

export default class Config implements ConfigInterface {
  extends: string | Array<string>;

  npmClient: NpmClients = 'npm';

  showConfigs: boolean;

  skills: Array<Skill>;

  autoInstall: boolean = false;

  static DEFAULT_CONFIG = {
    skills: [],
    showConfigs: false
  };

  constructor(config: UnresolvedConfigInterface | ResolvedConfigInterface | ConfigInterface) {
    ValidateConfig(config);
    this.normalizeWithResolvedSkills();

    this.extends = 'extends' in config ? config.extends : '';
    this.skills = config.skills;
    this.showConfigs = config.showConfigs;
    this.npmClient = config.npmClient;
  }

  getConfigWithDefaults(): UnresolvedConfigInterface {
    return {
      ...Config.DEFAULT_CONFIG,
      ...this.getConfigValues()
    };
  }

  getConfigValues(): UnresolvedConfigInterface {
    return {
      extends: this.extends,
      skills: this.skills,
      showConfigs: this.showConfigs,
      npmClient: this.npmClient,
      autoInstall: this.autoInstall
    };
  }

  /**
   * Write the config to a package.json file
   * @param pkgPath - The path to the package.json file
   */
  async write(pkgPath: string): Promise<string> {
    Project.validatePkgPath(pkgPath);
    return Config.writeToPkgJson(pkgPath, {
      ...JSON.parse((await fs.promises.readFile(pkgPath)).toString()),
      alfred: this.getConfigValues()
    });
  }

  private normalizeWithResolvedSkills(): ResolvedConfigInterface {
    const normalizedConfig = this.normalizeWithResolvedConfigs(this);
    if (!normalizedConfig.skills || !normalizedConfig.skills.length)
      return normalizedConfig;

    const skillsMap: ConfigMap = new Map();
    const mappedSkills: ConfigMap = normalizedConfig.skills.reduce(
      (map: ConfigMap, skill: ConfigSkill) => {
        if (typeof skill === 'string') {
          map.set(skill, {});
          return map;
        }
        if (Array.isArray(skill)) {
          const [skillName, skillConfig] = skill;
          if (map.has(skillName)) {
            map.set(
              skillName,
              mergeConfigs({}, map.get(skillName), skillConfig)
            );
          } else {
            map.set(skillName, skillConfig);
          }
          return map;
        }
        throw new Error(`Config type not supported: ${JSON.stringify(skill)}`);
      },
      skillsMap
    );

    normalizedConfig.skills = Array.from(mappedSkills.entries());

    return normalizedConfig;
  }

  /**
   * Initialize a config from the root directory of an alfred project
   */
  static initFromProjectRoot(projectRoot: string): Config {
    const pkgPath = path.join(projectRoot, 'package.json');
    Project.validatePkgPath(pkgPath);

    // Read the package.json and validate the Alfred config
    const { alfred } = JSON.parse(fs.readFileSync(pkgPath).toString());

    return new Config(alfred);
  }

  private normalizeWithResolvedConfigs(config: UnresolvedConfigInterface): ResolvedConfigInterface {
    if (!config.extends) return config;
    // Convert extends: 'my-config' to extends: ['my-config']
    if (typeof config.extends === 'string') {
      config.extends = [config.extends];
    }
    // If a config is a string, require it
    const normalizedConfigs = config.extends.map(_config =>
      typeof _config === 'string' ? requireConfig(_config) : _config
    );
    // If nothing to extend then return the config itself without the extends
    // property
    if (config.extends && !config.extends.length) {
      const newConfig = { ...config };
      delete newConfig.extends;
      return newConfig;
    }

    for (let i = 0; i < normalizedConfigs.length; i += 1) {
      // eslint-disable-next-line no-param-reassign
      normalizedConfigs[i] = this.normalizeWithResolvedConfigs(
        normalizedConfigs[i]
      );
    }

    const mergedConfig = mergeConfigs(
      {},
      ...normalizedConfigs.map(e => e),
      config
    );
    delete mergedConfig.extends;

    return mergedConfig;
  }

  /**
   * Merge an object to the existing package.json and write it
   */
  static async writeToPkgJson(pkgPath: string, obj: Object): Promise<string> {
    Project.validatePkgPath(pkgPath);
    const pkg = {
      ...JSON.parse((await fs.promises.readFile(pkgPath)).toString()),
      ...obj
    };
    const formattedPkg = await formatPkgJson(pkg);
    await fs.promises.writeFile(pkgPath, formattedPkg);
    return formattedPkg;
  }

  /**
   * @TODO Migrate to this API
   */
  generateCtf(interfaceState: InterfaceState) {
    // Generate the CTF
    const tmpCtf: CtfMap = new Map();
    const { skills = [] } = this;
    skills.forEach(([skillPkgName, skillConfig]) => {
      // Add the skill config to the ctfNode
      const ctfNode: CtfNode = require(skillPkgName);
      ctfNode.config = skillConfig;
      if (ctfNode.configFiles) {
        ctfNode.configFiles = ctfNode.configFiles.map(configFile => ({
          ...configFile,
          config: mergeConfigs(
            {},
            configFile.config,
            // Only apply config if skill has only one config file
            ctfNode.configFiles.length === 1 ? skillConfig : {}
          )
        }));
      }
      tmpCtf.set(ctfNode.name, ctfNode);
    });

    const ctf = CTF(Array.from(tmpCtf.values()), interfaceState);
    addMissingStdSkillsToCtf(this, ctf, interfaceState);

    return ctf;
  }
}
