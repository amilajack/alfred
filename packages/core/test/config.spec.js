import config, { getConfigs } from '../src/config';

describe('config', () => {
  describe('getConfig', () => {
    it('should take an object with ".extends" property', () => {
      expect(
        getConfigs({
          extends: [{ extends: [{ bar: 'zoo' }] }],
          bar: 'bar'
        })
      ).toEqual({
        bar: 'bar'
      });
      expect(
        getConfigs({
          extends: [{ extends: [{ bar: 'zoo' }, { bar: 'foo' }] }],
          bar: 'baz'
        })
      ).toEqual({
        bar: 'baz'
      });

      expect(
        getConfigs({
          extends: [{ extends: '' }]
        })
      ).toEqual({});
    });

    /**
     * @TODO
     */
    it.skip('should not allow recursive deps', () => {});
  });

  describe('config', () => {
    it('should take plain object', () => {
      expect(
        config({
          bar: 'bar'
        })
      ).toEqual({
        bar: 'bar'
      });
    });

    it('should take an object with empty extends', () => {
      expect(
        config({
          extends: [],
          bar: 'bar'
        })
      ).toEqual({
        bar: 'bar'
      });
    });

    it('should accept strings in ".extends" array', () => {
      jest.mock(
        'module-1',
        () => ({
          bar: 'bar',
          foo: 'foobar'
        }),
        { virtual: true }
      );
      jest.mock(
        'module-2',
        () => ({
          bar: 'baz',
          cow: 'bar',
          extends: 'module-1'
        }),
        { virtual: true }
      );
      expect(
        config({
          extends: 'module-1',
          bar: 'who',
          hello: 'jane'
        })
      ).toEqual({
        bar: 'who',
        foo: 'foobar',
        hello: 'jane'
      });
      expect(
        config({
          extends: ['module-2'],
          hello: 'john'
        })
      ).toEqual({
        bar: 'baz',
        cow: 'bar',
        foo: 'foobar',
        hello: 'john'
      });
    });

    it('should throw if extends property is not a string or an array', () => {
      expect(() =>
        config({
          extends: [{ extends: '' }]
        })
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        config({
          extends: () => {}
        })
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
