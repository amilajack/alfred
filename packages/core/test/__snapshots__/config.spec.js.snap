// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`config config should throw if extends property is not a string or an array 1`] = `"Values in \\".extends\\" property in Alfred config must be a string. Instead passed [{\\"extends\\":\\"\\"}]"`;

exports[`config config should throw if extends property is not a string or an array 2`] = `"Values in \\".extends\\" property in Alfred config must be a string. Instead passed undefined"`;
